src/client/opamAction.ml

```diff
@@ -366,15 +366,33 @@ let prepare_package_source st nv dir =
   in
   let check_extra_files =
     try
-      List.iter (fun (src,base,hash) ->
-          if not (OpamHash.check_file (OpamFilename.to_string src) hash) then
+      let extra_files_dir =
+        OpamPath.Switch.extra_files_dir st.switch_global.root st.switch
+      and metadata_dir = OpamFile.OPAM.metadata_dir opam
+      and extra_files =
+        OpamFile.OPAM.get_extra_files
+          ~repos_roots:(OpamRepositoryState.get_root st.switch_repos)
+          opam
+      and copy_if_hash_matches (src, base, hash) =
+          if OpamHash.check_file (OpamFilename.to_string src) hash then
+            let dst = OpamFilename.create dir base in
+            OpamFilename.copy ~src ~dst
+          else
             failwith
               (Printf.sprintf "Bad hash for %s" (OpamFilename.to_string src))
-          else
-            OpamFilename.copy ~src ~dst:(OpamFilename.create dir base))
-        (OpamFile.OPAM.get_extra_files
-           ~repos_roots:(OpamRepositoryState.get_root st.switch_repos)
-           opam);
+      in
+      (match metadata_dir with
+       | None ->
+         (* lookup in switch-local hashmap overlay *)
+         List.iter (fun (base, hash) ->
+             let src =
+               let base = OpamFilename.Base.of_string (OpamHash.contents hash) in
+               OpamFilename.create extra_files_dir base
+             in
+             copy_if_hash_matches (src, base, hash))
+           (match OpamFile.OPAM.extra_files opam with None -> [] | Some xs -> xs)
+       | Some _ ->
+         List.iter copy_if_hash_matches extra_files);
       None
     with e -> Some e
```

Comment:
I was thinking about something like that : first get the cached ones then the already present in files (downloaded) ones, and merge the two lists, giving priority to cached ones
```suggestion
  let check_extra_files =
    try
      let cached_files =
        match OpamFile.OPAM.extra_files opam with
        | None -> []
        | Some xs ->
          (* lookup in switch-local hashmap overlay *)
          let extra_files_dir =
            OpamPath.Switch.extra_files_dir st.switch_global.root st.switch
          in
          OpamStd.List.filter_map (fun (base, hash) ->
              let src =
                let base = OpamFilename.Base.of_string (OpamHash.contents hash) in
                OpamFilename.create extra_files_dir base
              in
              if OpamFilename.exists src then
                Some (src, base, hash)
              else None)
            xs
      in
      let files =
        match
          OpamFile.OPAM.get_extra_files
            ~repos_roots:(OpamRepositoryState.get_root st.switch_repos)
            opam with
        | [] -> cached_files
        | fs ->
          OpamStd.List.filter_map (fun (_, base, hash) ->
              List.find_opt (fun (_, base', hash') ->
                  base <> base' && hash <> hash')
                cached_files) fs
      in
      let bad_hash =
        OpamStd.List.filter_map (fun (src, base, hash) ->
            if OpamHash.check_file (OpamFilename.to_string src) hash then
              (OpamFilename.copy ~src ~dst:(OpamFilename.create dir base); None)
            else
              Some src) files
      in
      if bad_hash <> [] then
        failwith
          (Printf.sprintf "Bad hash for %s"
             (OpamStd.Format.pretty_list (List.map OpamFilename.to_string bad_hash)));
      None
    with e -> Some e
```

Reply:
hmm, as I tried to elaborate in my message, I don't like the semantics "take either from cache or from `files/`", but really prefer to take them _all_ from `files/` -- or if files is not present, take them all from cache.

Reply:
but if you think it is better (or makes it more likely to get merged) to take files from either here or there, feel free to push your changes directly on the branch.

Reply:
Oh ok. That behavior is ok for me too. cf. last commit

Reply:
thanks :)